resources/backend/sql_advanced.md

```diff
@@ -5,11 +5,78 @@
 
 ## What is an Index?
 ## What is a Constraint?
- - Not Null
- - Unique
- - Default
+[SQL constraints](https://www.w3schools.com/sql/sql_constraints.asp) are essentially rules that limit what data can be inserted into a table. Constraints are used to ensure that the data in a table is valid. This increases the reliability of data and allows people more control over their databases. 
+
+If a constraint is not met, SQL will throw an error and the action will be stopped.
+
+We will use the following `example` table to demonstrate each constraint in SQL:
+```mysql
+CREATE TABLE example
+(
+    id     int NOT NULL UNIQUE,
+    number int DEFAULT 0,
+    CHECK (number < 10),
+    PRIMARY KEY (id)
+);
+```
+
+#### PRIMARY KEY
+
+The PRIMARY KEY identifies each row in the table. Primary keys must be unique and cannot be null. In `example`, the PRIMARY KEY is `id`. A primary key can consist of one column or multiple columns.
+
+#### NOT NULL
+
+The NOT NULL constraint ensures that a column cannot have NULL values.
+
+Attempting to insert null into `id` yields: `ERROR 1048 (23000): Column 'id' cannot be null`.
+
+#### UNIQUE
+
+The UNIQUE constraint ensures that each value in the column is unique. The PRIMARY KEY constraint automatically makes `id` UNIQUE and NOT NULL. 
+Adding a duplicate value yields: `ERROR 1062 (23000): Duplicate entry '1' for key 'example.id'`
+
+#### DEFAULT
+
+The DEFAULT constraint specifies a value if one is not provided. Running `INSERT INTO example (id) VALUES (1);` will create a row with `id` having a value of 1 and `number` having a value of 0.
+
+#### CHECK
+
+The CHECK constraint only allows certain values in a column. In `example`, the check constraint ensures that `number` is less than 10.
+
+Violating the check constraint yields: `ERROR 3819 (HY000): Check constraint 'example_chk_1' is violated.`
+
+#### FOREIGN KEY
+
+See below
 
 ## What is a Foreign Key?
+
+Foreign keys links 2 tables together. A foreign key in one table refers to the primary key in another table. The FOREIGN KEY constraint prevents actions that would destroy the link.
+
+Foreign keys are useful because it allows us to establish complex relationships between tables. In real world databases, tables often refer to or depend on each other. Foreign keys allow us to accomplish this easily.
+
+Continuing the example from before, the `foreign_example` table below has a foreign key called `example_id` that refers to the `id` column in the `example` table.
+
+```mysql
+CREATE TABLE foreign_example
+(
+    id         int,
+    example_id int,
+    PRIMARY KEY (id),
+    FOREIGN KEY (example_id) REFERENCES example (id)
+);
+```
+
+A row can be inserted into `foreign_example` normally, but for the `example_id` column, the corresponding `id` should be used. Once a row is inserted, the `example` row it references cannot be updated or deleted. Attempting to do so yields 
```

Comment:
Add one more sentence about what would the correct way be to delete a row referencing a foreign key?

If you want to delete example.id 22, you must first delete all rows that reference to it (aka delete all rows in foreign_example that reference example.id 22). Once all of those are deleted, then you can delete example.id 22.

Reply:
Actually,  explain that concept after describing the below error messages

Reply:
You could also give examples of how we use foreign keys in our codebase (like what you did in the other file)